sw/device/lib/dif/dif_edn.c

```diff
@@ -219,18 +219,27 @@ dif_result_t dif_edn_get_main_state_machine(const dif_edn_t *edn,
 dif_result_t dif_edn_instantiate(
     const dif_edn_t *edn, dif_edn_entropy_src_toggle_t entropy_src_enable,
     const dif_edn_seed_material_t *seed_material) {
-  if (edn == NULL || seed_material == NULL) {
+  if (edn == NULL) {
     return kDifBadArg;
   }
+  dif_csrng_seed_material_t *seed_material_ptr;
   dif_csrng_seed_material_t seed_material2;
-  memcpy(&seed_material2, seed_material, sizeof(seed_material2));
+  if (seed_material == NULL) {
+    // Caller provided no seed material; simply pass the NULL pointer on.
+    seed_material_ptr = NULL;
+  } else {
+    // Caller provided seed material; copy it to the stack and pass a pointer
+    // to that.
+    memcpy(&seed_material2, seed_material, sizeof(seed_material2));
```

Comment:
What is the reason for this `memcpy()`? Why can't the `seed_material` pointer not be directly passed to `csrng_send_app_cmd()`, which will loop over the words in the contained array and individually write them to a register? Do we need to guarantee some alignment?

Reply:
That's not clear for me as well. I don't think the reason is alignment either.
There's unittests for this DIF so I think it is safe to refactor and check weather the tests will pass.

Reply:
I'm not sure the unittests would catch alignment problems. The reasoning is as follows: The `mmio_region_write32()` function requires values to be written to be 32-bit-aligned.  [C alignment requirements for `struct`s ensure that each member variable is self-aligned](http://www.catb.org/esr/structure-packing/#_structure_alignment_and_padding).  The `seed_material` pointer could be manipulated to violate this alignment, in which case the words in the seed array would also be unaligned.  This would violate the alignment requirement of     `mmio_region_write32()`.

To prevent this, the `dif_edn_instantiate()` function now checks that `seed_material` is aligned.

WDYT? Also, maybe @timothytrippel who reviewed the original commit, has further insights?

Reply:
> The seed_material pointer could be manipulated to violate this alignment

Of course in C you can do essentially everything. But, assuming you don't do anything that would be undefined behaviour, could you actually violate the alignment?

I agree that the `memcpy` call can be removed if we guarantee that either improperly aligned pointers are rejected or that they are not passed in the first place.

Reply:
> Of course in C you can do essentially everything. But, assuming you don't do anything that would be undefined behaviour, could you actually violate the alignment?

I don't think so, but since it's easy to let a pointer become misaligned by mistake, I think it's wise to a have a basic check against it.

> I agree that the `memcpy` call can be removed if we guarantee that either improperly aligned pointers are rejected or that they are not passed in the first place.

The current implementation rejects improperly aligned pointers before calling the function that does  `mmio_region_write32()`. It may be even better to move that check into the function doing the writes.

Reply:
> The current implementation rejects improperly aligned pointers before calling the function that does `mmio_region_write32()`. It may be even better to move that check into the function doing the writes. Do you agree?

In principle that sounds like a good idea but I don't know how practical it is. Would that be a check that the `offset` argument is divisible by 4? The function is marked as inline (which helps with optimizations) but I can imagine that it might still result in a significant code size increase. BTW, it would be nice to have a build configuration that would enable additional expensive checks, which would make it easier to add extra sanity checks without burdening too much the normal configurations.

Reply:
I was thinking of doing the checks in `csrng_send_app_cmd()` not `mmio_region_write32()` (that probably was not clear, sorry). So I think execution time or binary size should not be increased significantly (for non-degenerate code).

Reply:
Anyway I think it is worth adding a unittest checking alignment, see [kmac](https://github.com/lowRISC/opentitan/blob/a79ca638323fd1339a29f69763693fa8e920530e/sw/device/lib/dif/dif_kmac_unittest.cc#L573) as reference

Reply:
Thanks for your feedback :+1: I have implemented this as discussed, including the unit test. If you agree, this can be merged and I will create a separate PR to remove the `memcpy()` from `dif_edn_reseed()` and `dif_edn_update()` (and potentially allow `seed_material` to be `NULL` there as well, but I first have to check if that is possible).